ouroboros-network-framework/src/Simulation/Network/Snocket.hs

```diff
@@ -528,8 +536,92 @@ instance Show addr => Show (FD_ m addr) where
 
 -- | File descriptor type.
 --
-newtype FD m peerAddr = FD { fdVar :: (StrictTVar m (FD_ m peerAddr)) }
-
+newtype FD m peerAddr = FD { fdVar :: StrictTVar m (FD_ m peerAddr) }
+
+instance ( MonadST m
+         , MonadThrow m
+         , MonadSay m
+         , MonadLabelledSTM m
+         , Show addr
+         ) => ToRawBearer m (FD m (TestAddress addr)) where
+  toRawBearer = makeRawFDBearer
+
+-- | Make a 'RawBearer' from an 'FD'. Since this is only used for testing, we
+-- can bypass the requirement of moving raw bytes directly between file
+-- descriptors and provided memory buffers, and we can instead covertly use
+-- plain old 'ByteString' under the hood. This allows us to use the
+-- 'AttenuatedChannel' inside the `FD_`, even though its send and receive
+-- methods do not have the right format.
+makeRawFDBearer :: forall addr m.
+                   ( MonadST m
+                   , MonadLabelledSTM m
+                   , MonadThrow m
+                   , MonadSay m
+                   , Show addr
+                   )
+                => FD m (TestAddress addr)
+                -> m (RawBearer m)
+makeRawFDBearer (FD {fdVar}) = do
+    (bufVar :: StrictTMVar m LBS.ByteString) <- newTMVarIO LBS.empty
+    return RawBearer
+              { send = \src srcSize -> do
+                  labelTVarIO fdVar "sender"
+                  say $ "Sending " ++ show srcSize ++ " bytes"
```

Comment:
I would avoid that, because this can be very noisy.  It's better to provide a tracer which the user can fill with `nullTracer` or something better for debugging.

That's another argument to abandon the `ToRawBearer` type class (we probably don't want to have that tracer for real fds).

Reply:
Ah, but wait - this is the instance for the simulated snocket, which can only be used in `IOSim`. The "real" instances do not demand `MonadSay`, and they do not emit any trace logging at all. And `IOSim` itself already routes `say` through a `Tracer`, so this is actually working just fine - we get traces when running in `IOSim`, but no performance overhead when running in `IO`.

Reply:
Indeed, I've also thought of adding it to `Snocket`, but that would add a lot of noise to all snocket code, despite being a fairly niche feature - I expect that the only practical application of this will be for the KES Agent, where we need to send private keys over network sockets without ever leaking them to disk (including swap); anything else can just use the existing networking features and plain old serialization through intermediate values on the Haskell heap.

Your idea of injecting the tracer through `makeRawFDBearer` won't work, because the test code needs to run against both simulated and real snockets; since it's generalized over the `fd` type, it doesn't know to call `makeRawFDBearer`, so unless we give up on the generalization (which means we can no longer test the actual identical code against `IOSim` unchanged, which is the whole point of this), the only ways I can see for this to work would be:

1. Extend the `send` and `recv` methods to accept a `Tracer` as an extra argument. Downside is that we'll be passing tracers to production instances, which will then silently ignore it - a lot of noise, potentially misleading, and there might even be performance overhead to it as well.
2. Pass a function `fd -> m (RawBearer m)` around as an argument; we can now inject the tracer by closing that function over it. This is essentially the same as the typeclass approach, except that we're passing the typeclass dictionary (or the single method, in this case) around explicitly.
3. Same as 2., but make that function part of the `Snocket` data structure.

I'm not a huge fan of either, but if we really feel the need to change the tracer, then option 2 would be my preference.

Reply:
OK, ran with it and implemented a better tracing solution:
- got rid of the typeclass after all
- introduced `MakeRawBearer` type (analogous to `MakeBearer`)
- stringly-typed debug logging calls inside the simulated socket's RawBearer now use a structured trace message type, `FDRawBearerTrace`
- when making a `MakeRawBearer` for simulated sockets, caller must pass a `Tracer m FDRawBearerTrace`